lib/Backend/FlowGraph.cpp

```diff
@@ -95,13 +248,109 @@ FlowGraph::Build(void)
+            //
+            //          TryFinally L1
+            //          <try code>
+            //          BrOnException L1
+            //          Leave L1'
+            //  L1 :    BailOnException
+            //  L1':    Finally
+            //          <finally code>
+            //          LeaveNull
+            //  L3:     <code after try finally>
+            //We generate 2 flow edges from the try - an exception path and a non exception path.
+            //This transformation enables us to optimize on the non-excepting finally path
+
+            Assert(instr->m_prev->IsLabelInstr());
+
+            IR::LabelInstr * finallyLabel = instr->m_prev->AsLabelInstr();
+
+            // Find leave label
+
+            Assert(finallyLabel->m_prev->m_opcode == Js::OpCode::Br && finallyLabel->m_prev->m_prev->m_opcode == Js::OpCode::Label);
+
+            IR::Instr * insertPoint = finallyLabel->m_prev;
+            IR::LabelInstr * leaveTarget = finallyLabel->m_prev->m_prev->AsLabelInstr();
+
+            leaveTarget->Unlink();
+            finallyLabel->InsertBefore(leaveTarget);
+            finallyLabel->Unlink();
+            insertPoint->InsertBefore(finallyLabel);
+
+            // Bailout from the opcode following Finally
+            IR::Instr * bailOnException = IR::BailOutInstr::New(Js::OpCode::BailOnException, IR::BailOutOnException, instr->m_next, instr->m_func);
+            insertPoint->InsertBefore(bailOnException);
+            insertPoint->Remove();
+
+            this->finallyLabelStack->Push(finallyLabel);
+            Assert(!this->leaveNullLabelStack->Empty());
+            this->leaveNullLabelStack->Pop();
+
+            Assert(leaveTarget->labelRefs.HasOne());
+            IR::BranchInstr * brOnException = IR::BranchInstr::New(Js::OpCode::BrOnException, finallyLabel, instr->m_func);
+            leaveTarget->labelRefs.Head()->InsertBefore(brOnException);
+
+            instrPrev = instr->m_prev;
+        }
+        break;
+
         case Js::OpCode::TryCatch:
             if (this->catchLabelStack)
             {
                 this->catchLabelStack->Pop();
             }
             break;
 
+        case Js::OpCode::TryFinally:
+            if (this->finallyLabelStack)
+            {
+                this->finallyLabelStack->Pop();
+            }
+            break;
+        case Js::OpCode::LeaveNull:
+        {
+            if (!this->leaveNullLabelStack)
+            {
+                break;
+            }
+            IR::Instr * label = instr->GetPrevRealInstrOrLabel();
+            if (label->IsLabelInstr())
+            {
+                this->leaveNullLabelStack->Push(label->AsLabelInstr());
+                break;
+            }
+            // Create a new label before LeaveNull
+            IR::LabelInstr *leaveNullLabel = IR::LabelInstr::New(Js::OpCode::Label, this->func);
+            instr->InsertBefore(leaveNullLabel);
+            this->leaveNullLabelStack->Push(leaveNullLabel);
```

Comment:
then it might be worth asserting after the flowgraph is built that the leaveNullLabelStack is null or empty (similar to catchLabelStack and finallyLabelStack)